src/game.cpp

```diff
@@ -1789,10 +1789,83 @@ void game::update_weather()
     }
 }
 
+int game::get_heat_radiation( const tripoint &location, bool direct )
+{
+    // Direct heat from fire sources
+    // Cache fires to avoid scanning the map around us bp times
+    // Stored as intensity-distance pairs
+    int temp_mod = 0;
+    std::vector<std::pair<int, int>> fires;
+    fires.reserve( 13 * 13 );
+    int best_fire = 0;
+    for( const tripoint &dest : g->m.points_in_radius( location, 6 ) ) {
+        int heat_intensity = 0;
+
+        int ffire = g->m.get_field_strength( dest, fd_fire );
+        if( ffire > 0 ) {
+            heat_intensity = ffire;
+        } else if( g->m.tr_at( dest ).loadid == tr_lava ) {
+            heat_intensity = 3;
+        }
+        if( heat_intensity == 0 || !g->m.sees( location, dest, -1 ) ) {
+            // No heat source here
+            continue;
+        }
+        // Ensure fire_dist >= 1 to avoid divide-by-zero errors.
+        const int fire_dist = std::max( 1, square_dist( dest, location ) );
+        fires.emplace_back( std::make_pair( heat_intensity, fire_dist ) );
+        if( fire_dist <= 1 ) {
+            // Extend limbs/lean over a single adjacent fire to warm up
+            best_fire = std::max( best_fire, heat_intensity );
+        }
+    }
+
+    for( const auto &intensity_dist : fires ) {
+        const int intensity = intensity_dist.first;
+        const int distance = intensity_dist.second;
+        temp_mod += 6 * intensity * intensity / distance;
+    }
+    if( direct ) {
+        return best_fire;
+    }
+    return temp_mod;
+}
+
+int game::get_convection_temperature( const tripoint &location )
+{
+    // Heat from hot air (fields)
+    int temp_mod = 0;
+    const trap &trap_at_pos = g->m.tr_at( location );
+    // directly on fire/lava tiles
+    int tile_strength = g->m.get_field_strength( location, fd_fire );
+    if( tile_strength > 0 || trap_at_pos.loadid == tr_lava ) {
+        temp_mod += 300;
+    }
+    // hot air of a fire/lava
+    auto tile_strength_mod = []( const tripoint &loc, field_id fld, int case_1, int case_2, int case_3 ){
+        int strength = g->m.get_field_strength( loc, fld );
+        int cases[3] = { case_1, case_2, case_3 };
+        return ( strength > 0 && strength < 4 ) ? cases[ strength - 1 ] : 0;
+    };
+
+    temp_mod += tile_strength_mod( location, fd_hot_air1,  2,   6,  10 );
+    temp_mod += tile_strength_mod( location, fd_hot_air2,  6,  16,  20 );
+    temp_mod += tile_strength_mod( location, fd_hot_air3, 16,  40,  70 );
+    temp_mod += tile_strength_mod( location, fd_hot_air4, 70, 100, 160 );
+
+    return temp_mod;
+}
+
 int game::get_temperature( const tripoint &location )
 {
+    int temp_mod = 0; // local modifier
+    
+    if( !new_game ) {
+        temp_mod += g->get_heat_radiation( location, false );
+        temp_mod += g->get_convection_temperature( location );
```

Comment:
If `game::get_heat_radiation` and `game::get_convection_temperature` are made into free functions change calls to `get_heat_radiation( g->m, location, false );` and `get_convection_temperature( g->m, location );`